src/connector/src/source/pulsar/source/reader.rs

```diff
@@ -12,26 +12,92 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::HashMap;
 use std::time::{SystemTime, UNIX_EPOCH};
 
 use anyhow::{anyhow, ensure, Result};
+use arrow_array::{Int32Array, Int64Array, RecordBatch};
 use async_trait::async_trait;
 use futures::StreamExt;
 use futures_async_stream::try_stream;
+use icelake::catalog::{load_catalog, CATALOG_NAME, CATALOG_TYPE};
+use icelake::io::FileScanStream;
+use icelake::types::{Any, AnyValue, StructValueBuilder};
+use icelake::{Table, TableIdentifier};
 use itertools::Itertools;
 use pulsar::consumer::InitialPosition;
 use pulsar::message::proto::MessageIdData;
 use pulsar::{Consumer, ConsumerBuilder, ConsumerOptions, Pulsar, SubType, TokioExecutor};
+use risingwave_common::array::{DataChunk, StreamChunk};
+use risingwave_common::catalog::ROWID_PREFIX;
+use risingwave_common::error::RwError;
 
+use crate::aws_utils::{ACCESS_KEY, REGION, SECRET_ACCESS};
+use crate::error::ConnectorError;
 use crate::parser::ParserConfig;
 use crate::source::pulsar::split::PulsarSplit;
 use crate::source::pulsar::{PulsarEnumeratorOffset, PulsarProperties};
 use crate::source::{
     into_chunk_stream, BoxSourceWithStateStream, Column, CommonSplitReader, SourceContextRef,
-    SourceMessage, SplitId, SplitMetaData, SplitReader,
+    SourceMessage, SplitId, SplitMetaData, SplitReader, StreamChunkWithState,
 };
 
-pub struct PulsarSplitReader {
+pub enum PulsarSplitReader {
+    Broker(PulsarBrokerReader),
+    Iceberg(PulsarIcebergReader),
+}
```

Comment:
just a small confirmation, how to handle partition changes with topic offload?
For example, a persistent partitioned topic A, including `A-1` and `A-2`, some  of data within are offload to iceberg, how about increasing `A-3` atm?

Reply:
Iceberg has snapshots for each table modification, e.g. partition change, schema change. We can store snapshot id for first read, and use it for failure recovery later.

Reply:
no further question, thanks